### ISSUE TITLE
Extract type providers to standalone projects

### ISSUE THREAD

**Issue by enricosada** [Microsoft/visualfsharp#441]
move in-box type providers out of the box and turn into a community projects

Type providers:
- [ ] [DbmlFile](https://msdn.microsoft.com/en-us/library/hh362326.aspx)
    for a database schema encoded in a .dbml file.
- [ ] [EdmxFile](https://msdn.microsoft.com/en-us/library/hh362313.aspx)
    to access a database with the schema in an .edmx file, using a LINQ to Entities mapping.
- [ ] [ODataService](https://msdn.microsoft.com/en-us/library/hh362325.aspx)
   to access an OData service.
- [ ] [SqlDataConnection](https://msdn.microsoft.com/en-us/library/hh362320.aspx)
   to access a SQL database.
- [ ] [SqlEntityConnection](https://msdn.microsoft.com/en-us/library/hh362322.aspx)
   to access a database, using a LINQ to Entities mapping.
- [ ] [WsdlService](https://msdn.microsoft.com/en-us/library/hh362328.aspx)
   for a Web Services Description Language (WSDL) web service

### Open questions:

1. ownership: community, ms support
2. split FSharp.Data.TypeProviders in multiple packages?

### Ok

- [X] GAC assemblies: no :x: 

### Ideas

- [ ]  TPs rely on exe codegen tools, use new version of these tools
- [ ]  Use roslyn instead of csc to compile c# codegen tool output

**Comment by enricosada** [comment]
as a todo/roadmap issue, each one can be a separate issue

a good approach ihmo is:

- create a new repo (scaffold etc)
- add visualfsharp as git submodule
- import source/tests as linked item from submodule
- build  + tests + nuget
- when ok, pr to remove from codebase the code

pro:

- we can start without changing the original visualfsharp codebase
- is easy to apply/test if the visualfsharp codebase change


i can do the wsdlService when if we choose how to proceeed (vNext obv)

**Comment by dsyme** [comment]
It would be great to make progress on this.

I go back and forth about whether these should be one DLL/nuget-package or multiple.   The simplest path is probably still just to have one DLL (FSharp.Data.TypeProviders) and one nuget package.  I can see pros and cons both ways, so it is worth trying to assess what problems that would cause, e.g. 

PRO splitting:
* the OData and EF type providers will need to evolve, probably by switching away from using the command line code-generation tools. (The command line code-gen tools are,  AFAIK, not fully up-to-date with latest features available to C# clients, which call an API to do the code generation rather than calling a command-line tool).  This kind of evolution may best happen in split projects

CON splitting:

* much simpler to get the split done

* existing docs and samples always talk about FSharp.Data.TypeProviders.dll.  

* multiple repos need multiple active owners to manage PRs, testing etc.

Either way, it's important to know if the package(s) would be shipped and signed as Microsoft packages (with the corresponding support licences), or whether community-provided packages would be sufficient.

Cheers
don

**Comment by enricosada** [comment]
Yes, we need to balance the number of nuget/repo.

My random number is 2 :smile: 

We should group type providers **by dependencies**:

- **sql** (need reference to System.Data.dll): dbml, edmx, sql, sqlentity
  -  FSharp.Data.TypeProviders.Sql.dll ?
- **webservice client** (need reference to httpclient): wsdl and odata
  - FSharp.Data.TypeProviders.WebserviceClient.dll ?

two assemblies => two repositories.

- tests automation work is reused ( sql -> create database, http -> consume fake webservice )
- easier to manage multiple framework profiles
- 2 is a low number of repo :smile: not too much overhead

after the fork, each repository can choose to create multiple nuget packages, if needed.

This way we need only two repo (low number) and after the initial repo bootstrap ( wsdl + sqldataconnection ) is easier to add the others ( odata to http, dbml+edmx+sqlentity to sql) because test instrastructure (create database, create webservice) is already in place.

**Comment by enricosada** [comment]
Another question:

The visualfsharp extension should include these nugets (battery included standard library) in gac?

For example ruby standard library moved from an a single package to a group of gem (= nuget).
the [discussion](https://redmine.ruby-lang.org/issues/5481) and the result is https://github.com/rubysl (the ruby stdlib is big, like the .net bcl).
The equivalent f# stdlib is like fsharp.core nuget + fsharp.data.typeprovider nuget + etc.

Without assembly split, the ms package can include the FSharp.Data.TypeProviders.dll, downloaded from the  FSharp.Data.TypeProviders nuget, so backward compatibility.

I dont think it should (not for typeproviders, maybe for fsharp.core), but i ask anyway

**Comment by latkin** [comment]
We'd be happy to move these out.  I like the plan of new owners standing up new repo(s) and getting things in shape, then we pull the plug and remove from VF# deployment/tests.

Mini brain dump:
- At MVP summit we hear feedback that in-box Microsoft blessing is still very important at many companies, i.e. that moving this to pure community ownership could be harmful for adoption
- I think all of the TPs rely on exe codegen tools. Code for sniffing location of these tools only finds fairly old versions at this point.  Could cheaply get a lot of new support simply by updating to find the latest version of these tools.
- Some of the TPs take codegen tool output (in C#) and compile on the fly with csc.exe.  Maybe worth looking at Roslyn-based hosted compile?

**Comment by KevinRansom** [comment]
Please avoid relying on the GAC.  Our current dependence on it is a historic misstep, it won't really work cross platform anyway.  So please do not introduce any new GAC requirements.

**Comment by dsyme** [comment]
What GAC requirement is there? On FSharp.Core or something else?

**Comment by enricosada** [comment]
MS support is **really** important, this is a library already supported by MS, so is good to continue like that. Maybe this can be relaxed having a core team by community+ms?

I dont want GAC'es assemblies ( whoa, scary word :ghost: :ghost: ), just askin, i understand the answer is no :+1: 

the @latkin braindump should be the first issues on the new repo :smile: 

I'll add open question and decisions to the issue summary ( best place is the wiki, but is closed to collaboration, make it public! no spam :smile:  )

Anyway, this is future, but with we can start with wsdl (repo+submobule+build+ci -> nuget) without changing visualfsharp codebase.

I can do the bootstrap of this repo (like a poc), if someone ( /cc @dsyme @KevinRansom @latkin  ) help me. We can discuss on gitter/slack to move faster, after the open question are closed.

**Comment by dsyme** [comment]
Personally I feel that if Microsoft is planning to offer enterprise support including updated Microsoft-signed nuget packages for new functionality, then it's probably optimal just to keep these particular type providers in this repository and start accepting pull requests for them.  The testing is already in place here, for example, as are all the engineering guidelines, CLA mechanism, repository ownership, internal Microsoft packaging and engineering processes etc.  

If Microsoft plan to split out the packages into new repositories then it probably really has to be done by Microsoft repo owners. The new repo ownership would be by Microsoft anyway, and most of the cost will be in adjusting the Microsoft engineering, release and support processes.

Alternatively (or concomitantly) I think it's fine to fork out new "fsprojects" versions of the type providers to evolve the functionality in the F# community in the same way as many other F# community projects. This would likely enable a much faster iteration rate, for starters, a bit like has been done with the Visual F# Power Tools (though that's not a fork).

It's also a bit like FSharp.Core.  There is a general intention from everyone to move that into a nuget package, and indeed there is already a community/Xamarin-provided FSharp.Core nuget package.  But to get a Microsoft-published package, much the work is on the Microsoft internal side, e.g. to adjust packaging, templates, tooling.  Everyone wants to help with this but there's a limit to what can be done without visibility of the internal Microsoft build and release processes. And there are limits w.r.t. timing too (Visual Studio 2015 v. vNext v. vNext++)

Anyway, I don't think there's any particular rush with this.  Unblocking it would be good but it may be one of those things we just need to wait for.

**Comment by enricosada** [comment]
Ok got it, thx @dsyme for feedback, we need to wait ms decision about support first.

i understand now is ok having the code inside visualfsharp repo, but moving to a new repo  can help to have an **independent versioning** ( that's really important because visualfsharp is really more locked down than a library like wsdl type provider, more easier to evolve/fix ),  decrease the visualfsharp codebase (less code/tests), speedup things in visualfsharp (tests suite)

i was trying to bootstrap the repo ( without a proper fork, there is already visualfsharp/openfsharp/FSharp.Compiler.Service and is good to have only a crossplat project, not crossplat vs ms ), but i am too hasty :smile: 

It's only that i see https://github.com/dotnet projects and seem an easier flow, but obviously is not, sry

**Comment by dsyme** [comment]
The code for FSharp.Data.TypeProviders is now available as an F# community library and nuget package at https://github.com/fsprojects/FSharp.Data.TypeProviders/. I'll close this issue and move it over to there, since splitting the type providers is still a plausible thing to do (though I suspect there's no particular reason to do it at the moment)

**Comment by enricosada** [comment]
Really thx @dsyme  . 
That's good because make it easier to contribute. It's a really nice thing. :+1: 

About split i'll comment in the other repo. 
I think only wsdl it's usefull atm, other tp are deprecated by community projects ( more advanced ). 
tp in this repo are pretty much dead code now ( a good thing )
